test/e2e/e2e.go

```diff
@@ -1491,6 +1502,122 @@ var _ = ginkgo.Describe("e2e multiple ecmp external gateway validation", func()
+			args := []string{
+				"netexec",
+				fmt.Sprintf("--http-port=%d", endpointHTTPPort),
+				fmt.Sprintf("--udp-port=%d", endpointUDPPort),
+			}
+			pod := createPod(f, node.Name+"-ep", node.Name, f.Namespace.Name, []string{}, endpointsSelector, func(p *v1.Pod) {
+				p.Spec.Containers[0].Args = args
+			})
+			endPoints = append(endPoints, pod)
+			nodesHostnames.Insert(pod.Name)
+
+			// this is arbitrary and mutuated from k8s network e2e tests. We aim to hit all the endpoints at least once
+			maxTries = len(endPoints)*len(endPoints) + 30
+		}
+
+		ginkgo.By("Creating the nodeport service")
+		npSpec := nodePortServiceSpecFrom(nodePortServiceName, endpointHTTPPort, endpointUDPPort, clusterHTTPPort, clusterUDPPort, endpointsSelector)
+		np, err := f.ClientSet.CoreV1().Services(f.Namespace.Name).Create(context.Background(), npSpec, metav1.CreateOptions{})
+		nodeTCPPort, nodeUDPPort = nodePortsFromService(np)
+		framework.ExpectNoError(err)
+
+		ginkgo.By("Waiting for the endpoints to pop up")
+		err = framework.WaitForServiceEndpointsNum(f.ClientSet, f.Namespace.Name, nodePortServiceName, len(endPoints), time.Second, wait.ForeverTestTimeout)
+		framework.ExpectNoError(err, "failed to validate endpoints for service %s in namespace: %s", nodePortServiceName, f.Namespace.Name)
+
+		ginkgo.By("Creating an external container to send the traffic from")
+		// the client uses the netexec command from the agnhost image, which is able to receive commands for poking other
+		// addresses.
+		createClusterExternalContainer(clientContainerName, agnhostImage, []string{"--network", "kind", "-P"}, []string{"netexec", "--http-port=80"})
+	})
+
+	ginkgo.AfterEach(func() {
+		deleteClusterExternalContainer(clientContainerName)
+	})
+
+	ginkgo.It("Should provide external connectivity via nodeport", func() {
+		for _, protocol := range []string{"http", "udp"} {
+			for _, node := range nodes.Items {
+				for _, nodeAddress := range node.Status.Addresses {
+					// skipping hostnames
+					addr := net.ParseIP(nodeAddress.Address)
+					if addr == nil {
+						continue
+					}
+
+					responses := sets.NewString()
+					valid := false
+					ginkgo.By("Hitting the nodeport on " + node.Name + " and reaching all the endpoints " + protocol)
+					for i := 0; i < maxTries; i++ {
+						nodePort := nodeTCPPort
+						if protocol == "udp" {
+							nodePort = nodeUDPPort
+						}
+						epHostname := pokeEndpointHostname(clientContainerName, protocol, nodeAddress.Address, nodePort)
+						responses.Insert(epHostname)
+
+						// each endpoint returns its hostname. By doing this, we validate that each ep was reached at least once.
+						if responses.Equal(nodesHostnames) {
+							framework.Logf("Validated node %s on address %s after %d tries", node.Name, nodeAddress.Address, i)
+							valid = true
+							break
+						}
+					}
+					framework.ExpectEqual(valid, true, "Validation failed for node", node, responses, nodeUDPPort)
```

Comment:
this should be:
framework.ExpectEqual(valid, true, "Validation failed for node", node, responses, nodePort)

Reply:
good catch!